Add clear button to search fields

Tunous opened the pull request:
This pull request adds functionality of clear **button** located inside of LineEdit (where right_icon is located), which **when clicked removes whole typed text**. It can be enabled similarly to how it's done for right_icon, by calling `add_icon_override("clear_text_icon", <icon>)`.

For now I've added it to all input fields that have right_icon defined, but this can be further extended by adding it where needed.

## Preview
![2018-07-15_15-45-34](https://user-images.githubusercontent.com/5156340/42734571-a254dc6e-8846-11e8-8472-ff6e31564704.gif)

## Issues

- [x] Positioning

Right now positioning doesn't work correctly for input fields that have alignment other than fill or left. This is because right_icon appears to be broken for these on master branch too and fixing it would be outside of the scope of this pull request. I can try to fix this if needed.

- [x] ~~Click state~~

~~I'm still new to the codebase so I didn't find how it could be implemented yet, but ideally the button should react to hover and click states by changing its colors the same way how ToolButton does it.~~

reduz commented:
This is interesting and PR seems ok. Please make sure it works with all alignments so we can merge.

Tunous commented:
Done, I've fixed positioning and rebased commits onto master. This is how it looks now with icon enabled or disabled:

![image](https://user-images.githubusercontent.com/5156340/43034319-972b6eb4-8cda-11e8-95c4-ec855853ee3a.png)

I've also added this button to few more places. Here is a full list now:

- Project Manager - project filter
- AssetLib - search field
- FileSystem dock - files search field
- Inspector dock - properties filter
- Scene dock - nodes filter
- Resource creation dialog - search field
- Group Editor - filter fields
- Quick Open dialogs - search field
- Documentation search dialog - search field
- Classes search dialog - search field
- TileMap - tile filter
- Editor Settings - search field in both tabs
- Project Settings - search field
- Visual Script editor - nodes filter

akien-mga commented:
Could you squash the commits together? It would be easier to review with a commit implementing it all than the various commits with iterative development and fixups.

akien-mga commented on scene/gui/line_edit.h:
```diff
@@ -150,6 +157,8 @@ class LineEdit : public Control {
 	virtual bool can_drop_data(const Point2 &p_point, const Variant &p_data) const;
 	virtual void drop_data(const Point2 &p_point, const Variant &p_data);
 
+	virtual Control::CursorShape get_cursor_shape(const Point2 &p_pos) const;
```
`LineEdit` inherits `Control`, so you might be able to use `CursorShape` directly without the namespace, no?

Tunous commented:
Done.

akien-mga commented:
This should implemented without touching `LineEdit` at all. Currently this is pretty hacky to check if there's an editor-specific icon defined to enable some specific behaviour.

It would be better to make a new EditorSearchBox control that extends LineEdit and has the proper icons added, and the logic to handle their action.

reduz commented:
I am fine with this being on LineEdit, because the feature is usually needed for mobile.
That said, there should be default icons and colors in scene/resources/default_theme.cpp and the images there.. then the editor can simply override them from the theme, there is no need to do all this custom override code.

reduz commented:
turning the icon on/off should simply be a bool property of LineEdit, not depending on it being in the theme

Tunous commented:
> make a new EditorSearchBox control that extends LineEdit

I can do that.

> has the proper icons added

Should `add_icon_override` still be used for that or is there another way which would be preferred to set these icons?

Tunous committed:
Add clear text button to LineEdit

- Add pressed state to clear button
- Enable clear button on all inputs with search icon
- Remove duplicate clear buttons
- Fix rendering of icon for center and right alignments
- Add clear button to more search fields
- Add clear icon to default theme
- Add method to control enabled state of clear button
- Add property to enable clear button from inspector

Tunous commented:
I've updated the code based on your comments. The clear button can now be enabled or disabled using `set_clear_button_enabled(bool)` method. I've also added corresponding property to the inspector as seen below. LineEdit in scene view automatically updates to reflect the changes.

![2018-07-26_13-15-42](https://user-images.githubusercontent.com/5156340/43259322-0b1a172c-90d6-11e8-9aaa-3094dd0fdf67.gif)

reduz commented on editor/project_settings_editor.cpp:
```diff
@@ -119,7 +120,8 @@ void ProjectSettingsEditor::_notification(int p_what) {
 
 			search_button->set_icon(get_icon("Search", "EditorIcons"));
-			clear_button->set_icon(get_icon("Close", "EditorIcons"));
+			search_box->add_icon_override("right_icon", get_icon("Search", "EditorIcons"));
```
This is not needed, you can override the icon directly here in the editor theme:

https://github.com/godotengine/godot/blob/master/editor/editor_themes.cpp#L827

Tunous replied:
Wouldn't that override the icon for all LineEdit's? Here I implemented it the same way to how it was done previously in other places.

reduz replied:
Yes, it is correct to override the icon for all line edits, since it's part of the editor theme (which overrides most of the default theme). If you want to enable clearing, just enable it via function.

Tunous replied:
I did it like this because that's how it was done previously. (For example [here](https://github.com/godotengine/godot/blob/master/editor/editor_help.cpp#L256)). Anyway, as requested I've changed all these calls to `set_right_icon function` which doesn't use theme.

Tunous committed:
Do not use theme to set LineEdit right_icon

akien-mga commented:
Thanks!
